### PR TITLE
Feat: Add 'thread' field to messages.

### PR DESCRIPTION
**Pull request by Artemis21**
## Summary

Parse and use the `thread` field on messages, as documented [here](https://discord.com/developers/docs/resources/channel#message-object).

## Checklist

- If code changes were made then they have been tested.
- I have updated the documentation to reflect the changes.
- This PR adds something new (e.g. new method or parameters).

### CODE REVIEW

**Review by ooliver1 — APPROVED** [review]
I assume this has been tested

**Comment by Artemis21** [comment]
Yes, I made the same change to a Discord.py fork a while back and it has been used successfully in production.

File: nextcord/message.py [hunk 1]
```diff
@@ -677,6 +682,12 @@ def __init__(
         except AttributeError:
             self.guild = state._get_guild(utils._get_as_snowflake(data, 'guild_id'))
 
+        thread_data = data.get('thread')
+        if self.guild and thread_data:
```
Comment (TAG-Epic) [inline nextcord/message.py:686]: Threads are a guild only feature? Why do this check?

**Review by TAG-Epic — CHANGES_REQUESTED** [review]
Needs testing too, don't have time atm

File: nextcord/message.py (hunk 1 above)
Comment (Artemis21) [reply nextcord/message.py:686]: To convince the type checker (and myself) that it won't break. I can remove it if you want, but I think it's best for safety - Discord may theoretically add threads to DMs in the future, in which case it's better that this code doesn't load them than that it breaks.

**Comment by Artemis21** [comment]
I have tested it, and the branch this pull request comes from is currently being used in production (and actively using the feature) without issues.

File: nextcord/message.py (hunk 1 above)
Comment (TAG-Epic) [reply nextcord/message.py:686]: Honestly it should just break... Doubt they will be adding thread support for DMs

File: nextcord/message.py [hunk 2]
```diff
@@ -597,6 +597,10 @@ class Message(Hashable):
     components: List[:class:`Component`]
         A list of components in the message.
 
+        .. versionadded:: 2.0
+    thread: Optional[:class:`Thread`]
+        The thread created from this message, if any.
```
Comment (RPSMain) [inline nextcord/message.py:602]:
```suggestion
        The thread created from a message, if any.
```

**Review by RPSMain — APPROVED** [review]

**Comment by Skelmis** [comment]
@Artemis21 Any updates?

**Comment by Artemis21** [comment]
@Skelmis addressed reviews...

**Review by vcokltfre — APPROVED** [review]

**Review by TAG-Epic — APPROVED** [review]
LGTM, not tested

**Comment by koala9712** [comment]
Returns `None` in a regular channel as well as a thread
with code:
```python
@commands.command()
async def test(self, ctx):
    await ctx.send(str(ctx.message.thread))
```

image
image

**Comment by Artemis21** [comment]
@koala9712 

> Returns `None` in a regular channel as well as a thread with code:

To get the channel or thread a message was sent in, use `Message.channel`. This attribute is for determining which if any thread was created *from* this message, as documented.

**Comment by koala9712** [comment]
>This attribute is for determining which if any thread was created _from_ this message, as documented.

So... this should theoretically have...?
unknown
unknown

**Pull request merged**